lib/ansible/plugins/inventory/foreman.py

```diff
@@ -121,134 +126,139 @@ def _get_session(self):
-            while True:
-                ret = s.get(url, params=params)
-                if ignore_errors and ret.status_code in ignore_errors:
-                    break
-                ret.raise_for_status()
-                json = ret.json()
-
-                # process results
-                # FIXME: This assumes 'return type' matches a specific query,
-                #        it will break if we expand the queries and they dont have different types
-                if 'results' not in json:
-                    # /hosts/:id dos not have a 'results' key
-                    results = json
+            if inventory is not None:
+                return inventory
+
+        params = {'page': 1, 'per_page': 250}
+
+        if self.get_option('want_params'):
+            params['include'] = 'all_parameters'
+
+        results = []
+        s = self._get_session()
+
+        while True:
+            ret = s.get("%s/api/v2/hosts" % self.foreman_url, params=params)
+            ret.raise_for_status()
+            json = ret.json()
+
+            # process results
+            # FIXME: This assumes 'return type' matches a specific query,
+            #        it will break if we expand the queries and they dont have different types
+            if 'results' not in json:
+                # /hosts/:id dos not have a 'results' key
+                results = json
+                break
+            elif isinstance(json['results'], MutableMapping):
+                # /facts are returned as dict in 'results'
+                results = json['results']
+                break
+            else:
+                # /hosts 's 'results' is a list of all hosts, returned is paginated
+                results = results + json['results']
+
+                # check for end of paging
+                if len(results) >= json['subtotal']:
                     break
-                elif isinstance(json['results'], MutableMapping):
-                    # /facts are returned as dict in 'results'
-                    results = json['results']
+                if len(json['results']) == 0:
+                    self.display.warning("Did not make any progress during loop. expected %d got %d" % (json['subtotal'], len(results)))
                     break
-                else:
-                    # /hosts 's 'results' is a list of all hosts, returned is paginated
-                    results = results + json['results']
 
-                    # check for end of paging
-                    if len(results) >= json['subtotal']:
-                        break
-                    if len(json['results']) == 0:
-                        self.display.warning("Did not make any progress during loop. expected %d got %d" % (json['subtotal'], len(results)))
-                        break
-
-                    # get next page
-                    params['page'] += 1
+                # get next page
+                params['page'] += 1
 
+        if self.use_cache:
```

Comment:
Right now the cache isn't set if a `- meta: refresh_inventory` task happens in a playbook because self.use_cache will be False. In parse() you could add a self.reset_cache boolean in parse():
```
# Both of these should be True when we want to read the cache
self.use_cache = cache and self.get_option('cache')

# cache is False when the inventory is being reset
self.reset_cache = not cache and self.get_option('cache')
```

Then this line should be `if self.use_cache or self.reset_cache:` so we set the case in both cases.

Reply:
Alternatively, you can just always set the cache like it was doing before, because if self.get_option('cache') is not True, then it's just an in-memory cache and no long-term caching will occur.